src/server/detail/save_stages_controller.h

```diff
@@ -19,16 +17,28 @@ class Service;
 
 namespace detail {
 
+class SnapshotStorage;
+
+struct LastSaveInfo {
+  // last success save info
+  time_t save_time = 0;  // epoch time in seconds.
+  uint32_t success_duration_sec = 0;
+  std::string file_name;                                      //
+  std::vector<std::pair<std::string_view, size_t>> freq_map;  // RDB_TYPE_xxx -> count mapping.
+  // last error save info
+  GenericError last_error;
+  time_t last_error_time = 0;      // epoch time in seconds.
+  time_t failed_duration_sec = 0;  // epoch time in seconds.
+};
+
 struct SaveStagesInputs {
   bool use_dfs_format_;
   std::string_view basename_;
   Transaction* trans_;
   Service* service_;
-  std::atomic_bool* is_saving_;
   util::fb2::FiberQueueThreadPool* fq_threadpool_;
   LastSaveInfo* last_save_info_ ABSL_GUARDED_BY(save_mu_);
   util::fb2::Mutex* save_mu_;
```

Comment:
Just a though I realize now (and not while extracting this class), it'd be more convenient if returned LastSaveInfo just by value from `Save()` instead of passing a mutex by pointer